Panache repository breaks when narrowing the generic of a subclass

### Describe the bug

I am not 100% sure what is going on here. I suspect it is a bug in Quarkus but I can't be 100% sure (the error message is definitely not correct though).

We use a setup where we have our own generic panache base repository, which inherits/extends the Quarkus provided panache base repository, and is then implemented by concrete repositories. A logical implementation of the intermediate layer is:
```kotlin
abstract class BaseRepository<EntityT : Any, IdT: Any> : PanacheRepositoryBase<EntityT, IdT>
```

Which indeed works perfectly fine. However, we now need something like this:
```kotlin
abstract class BaseRepository<EntityT : Any, IdT: UUIDWrapper> : PanacheRepositoryBase<EntityT, UUID>
```

And this suddenly breaks, with:
```
Caused by: java.lang.IllegalStateException: This method is normally automatically overridden in subclasses: did you forget to annotate your entity with @Entity?
	at io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations.implementationInjectionMissing(AbstractJpaOperations.java:452)
	at io.quarkus.hibernate.orm.panache.kotlin.PanacheRepositoryBase.findById(PanacheRepositoryBase.kt:95)
	at nl.rug.digitallab.BrokenBaseRepository.findById(Reproducer.kt:36)
	at nl.rug.digitallab.BrokenBaseRepository.findById(Reproducer.kt:36)
	at nl.rug.digitallab.BrokenBaseRepository.findByIdOrThrow(Reproducer.kt:40)
	at nl.rug.digitallab.BrokenAssignmentRepository.findByIdOrThrow(Reproducer.kt:46)
	at nl.rug.digitallab.BrokenAssignmentRepository_ClientProxy.findByIdOrThrow(Unknown Source)
	at Reproducer.This breaks for some reason(Reproducer.kt:32)
```

Even though we did not change the entity at all.

I tried to make a toy reproducer that has the minimal number of changes with the two to see where it breaks, and taking the exact same implementation, but changing
```kotlin
abstract class BaseRepository<EntityT : Any, IdT: Any> : PanacheRepositoryBase<EntityT, IdT>
```

by this (with making no other changes):

```kotlin
abstract class BaseRepository<EntityT : Any, IdT: UUID> : PanacheRepositoryBase<EntityT, IdT>
```

Suddenly makes this issue appear.

### Expected behavior

I should be able to narrow down the generic from Any to UUID without causing an issue with the generated entity method

### Actual behavior

An error is thrown when I narrow down the generic of my own base class

### How to Reproduce?

I have created a minimal reproducer. The steps are:
1. Clone https://gitlab.com/l.s.andringa1/quarkus-panache-reproducer.git
2. Run `gradle test`. I have added a test using a working repository, and a test for the broken issue (where the only difference is the narrowed generic).

### Output of `uname -a` or `ver`

Ubuntu 25.04

### Output of `java -version`

Java 21.0.8

### Quarkus version or git rev

3.26.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 9.0.0

### Additional information

_No response_

--- a/extensions/panache/hibernate-orm-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/orm/panache/kotlin/deployment/test/Bug50144Test.kt
+++ b/extensions/panache/hibernate-orm-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/orm/panache/kotlin/deployment/test/Bug50144Test.kt
@@ -1,0 +1,100 @@
+package io.quarkus.hibernate.orm.panache.kotlin.deployment.test
+
+import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepositoryBase
+import io.quarkus.test.QuarkusUnitTest
+import jakarta.enterprise.context.ApplicationScoped
+import jakarta.inject.Inject
+import jakarta.persistence.Entity
+import jakarta.persistence.GeneratedValue
+import jakarta.persistence.GenerationType
+import jakarta.persistence.Id
+import jakarta.transaction.Transactional
+import org.jboss.shrinkwrap.api.ShrinkWrap
+import org.jboss.shrinkwrap.api.spec.JavaArchive
+import org.junit.jupiter.api.extension.RegisterExtension
+import org.junit.jupiter.api.assertThrows
+import org.junit.jupiter.api.Test
+import java.util.UUID
+
+class Bug50144Test {
+    companion object {
+        @RegisterExtension
+        @JvmField
+        var runner = QuarkusUnitTest()
+            .setArchiveProducer {
+                ShrinkWrap.create(JavaArchive::class.java)
+                    .addClasses(WorkingAssignmentRepository::class.java, BrokenAssignmentRepository::class.java,
+                        WorkingBaseRepository::class.java, BrokenBaseRepository::class.java,
+                        Assignment::class.java, NotFoundException::class.java)
+            }
+    }
+
+    @Inject
+    private lateinit var workingAssignmentRepository: WorkingAssignmentRepository
+
+    @Inject
+    private lateinit var brokenAssignmentRepository: BrokenAssignmentRepository
+
+    @Transactional
+    @Test
+    fun `A working example`() {
+        val id: UUID = UUID.randomUUID()
+
+        assertThrows<NotFoundException> {
+            workingAssignmentRepository.findByIdOrThrow(id)
+        }
+    }
+
+    @Transactional
+    @Test
+    fun `This breaks for some reason`() {
+        val id: UUID = UUID.randomUUID()
+
+        assertThrows<NotFoundException> {
+            brokenAssignmentRepository.findByIdOrThrow(id)
+        }
+    }
+
+    class NotFoundException: Exception()
+
+    @Entity
+    class Assignment {
+        @Id
+        @GeneratedValue(strategy = GenerationType.UUID)
+        private lateinit var _id: UUID
+    }
+
+    abstract class WorkingBaseRepository<EntityT : Any, IdT: Any> : PanacheRepositoryBase<EntityT, IdT> {
+        fun findByIdOrThrow(
+            id: IdT,
+            exception: (IdT) -> Exception,
+        ): EntityT = findById(id) ?: throw exception(id)
+    }
+
+    @ApplicationScoped
+    class WorkingAssignmentRepository : WorkingBaseRepository<Assignment, UUID>() {
+        fun findByIdOrThrow(assignmentId: UUID): Assignment {
+            return findByIdOrThrow(assignmentId) {
+                NotFoundException()
+            }
+        }
+    }
+
+    abstract class BrokenBaseRepository<EntityT : Any, IdT: UUID> : PanacheRepositoryBase<EntityT, IdT> {
+        fun findByIdOrThrow(
+            id: IdT,
+            exception: (IdT) -> Exception,
+        ): EntityT = findById(id) ?: throw exception(id)
+    }
+
+    @ApplicationScoped
+    class BrokenAssignmentRepository : BrokenBaseRepository<Assignment, UUID>() {
+        fun findByIdOrThrow(assignmentId: UUID): Assignment {
+            return findByIdOrThrow(assignmentId) {
+                NotFoundException()
+            }
+        }
+    }
+
+
+}

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/KotlinPanacheClassOperationGenerationVisitor.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/KotlinPanacheClassOperationGenerationVisitor.java
@@ -4,7 +4,6 @@ import static io.quarkus.deployment.util.AsmUtil.getLoadOpcode;
 import static io.quarkus.deployment.util.AsmUtil.unboxIfRequired;
 import static io.quarkus.gizmo.Gizmo.ASM_API_VERSION;
 import static io.quarkus.panache.common.deployment.PanacheConstants.DOTNAME_GENERATE_BRIDGE;
-import static java.lang.String.format;
 import static java.util.stream.Collectors.toList;
 import static org.objectweb.asm.Opcodes.ACC_BRIDGE;
 import static org.objectweb.asm.Opcodes.ARRAYLENGTH;
@@ -25,10 +24,11 @@ import static org.objectweb.asm.Type.getType;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.StringJoiner;
-import java.util.TreeMap;
 import java.util.function.Function;
 
 import org.jboss.jandex.AnnotationInstance;
@@ -68,10 +68,11 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
     protected final Function<String, Type> argMapper;
     protected final ClassInfo classInfo;
     protected final ByteCodeType entityUpperBound;
+    // These are type arguments to the Panache base entity/repo, not to the current or intermediate type
     protected final Map<String, ByteCodeType> typeArguments = new HashMap<>();
     private final ByteCodeType baseType;
-    private final Map<String, MethodInfo> definedMethods = new TreeMap<>();
-
+    private final Set<String> userMethods = new HashSet<>();
+    private final Set<String> baseTypeMethods = new HashSet<>();
     private final Map<String, String> erasures = new HashMap<>();
     private final IndexView indexView;
     protected List<PanacheMethodCustomizer> methodCustomizers;
@@ -101,10 +102,20 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
                     ? byteCodeType.get()
                     : null;
         };
+        loadBaseTypeMethods();
+    }
 
-        collectMethods(classInfo);
-        filterNonOverrides();
-
+    /**
+     * This loads the signatures of every base type method that requires a bridge
+     */
+    private void loadBaseTypeMethods() {
+        for (MethodInfo method : indexView.getClassByName(baseType.dotName()).methods()) {
+            String descriptor = method.descriptor(type -> typeArguments.getOrDefault(type, OBJECT).get());
+            AnnotationInstance bridge = method.annotation(DOTNAME_GENERATE_BRIDGE);
+            if (bridge != null) {
+                baseTypeMethods.add(method.name() + "/" + descriptor);
+            }
+        }
     }
 
     public static List<ByteCodeType> recursivelyFindEntityTypeArguments(IndexView indexView, DotName clazz,
@@ -217,27 +228,6 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
         }
     }
 
-    private void collectMethods(ClassInfo classInfo) {
-        if (classInfo != null && !classInfo.name().equals(baseType.dotName())) {
-            classInfo.methods()
-                    .forEach(method -> {
-                        String descriptor = method.descriptor(m -> {
-                            ByteCodeType byteCodeType = typeArguments.get(m);
-                            return byteCodeType != null ? byteCodeType.get() : OBJECT.get();
-                        });
-                        MethodInfo prior = definedMethods.put(method.name() + descriptor, method);
-                        if (prior != null && !isBridgeMethod(method)) {
-                            throw new IllegalStateException(format("Should not run in to duplicate " +
-                                    "mappings: \n\t%s\n\t%s\n\t%s", method, descriptor, prior));
-                        }
-                    });
-            DotName superName = classInfo.superName();
-            if (superName != null) {
-                collectMethods(indexView.getClassByName(superName));
-            }
-        }
-    }
-
     private String desc(String name) {
         String s = name.replace(".", "/");
         return s.startsWith("[") ? s : "L" + s + ";";
@@ -311,16 +301,6 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
 
     private Label endLabel() {
         return labels.get(labels.size() - 1);
-    }
-
-    private void filterNonOverrides() {
-        new ArrayList<>(definedMethods.values())
-                .forEach(method -> {
-                    AnnotationInstance generateBridge = method.annotation(DOTNAME_GENERATE_BRIDGE);
-                    if (generateBridge != null) {
-                        definedMethods.remove(method.name() + method.descriptor());
-                    }
-                });
     }
 
     private void generate(MethodInfo method) {
@@ -460,10 +440,6 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
         mv.visitInsn(AsmUtil.getReturnInstruction(returnType));
     }
 
-    private boolean isBridgeMethod(MethodInfo method) {
-        return (method.flags() & ACC_BRIDGE) != ACC_BRIDGE;
-    }
-
     private org.objectweb.asm.Type asmType(Type methodParameter) {
         org.objectweb.asm.Type parameter;
         if (methodParameter.kind() == Type.Kind.TYPE_VARIABLE) {
@@ -531,23 +507,15 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
     @Override
     public MethodVisitor visitMethod(int access, String name, String descriptor, String signature,
             String[] exceptions) {
-
-        MethodInfo methodInfo = definedMethods.entrySet().stream()
-                .filter(e -> e.getKey().equals(name + descriptor))
-                .map(e -> e.getValue())
-                .findFirst()
-                .orElse(null);
-        if (methodInfo != null && !methodInfo.hasAnnotation(DOTNAME_GENERATE_BRIDGE)) {
-            return super.visitMethod(access, name, descriptor, signature, exceptions);
-        } else if (name.contains("$")) {
-            //some agents such as jacoco add new methods, they generally have $ in the name
-            return super.visitMethod(access, name, descriptor, signature, exceptions);
-        } else if (name.equals(CTOR_METHOD_NAME) || name.equals(CLINIT_METHOD_NAME)) {
-            //Arc can add no-args constructors to support intercepted beans
-            // Logging with Panache can add a class initializer
-            return super.visitMethod(access, name, descriptor, signature, exceptions);
+        // Kotlinc or something will add bridge methods for the base type methods, these are not user methods
+        // so we filter them out since we add them back in visitEnd()
+        String sig = name + "/" + descriptor;
+        if ((access & Opcodes.ACC_BRIDGE) != 0
+                && baseTypeMethods.contains(sig)) {
+            return null;
         }
-        return null;
+        userMethods.add(sig);
+        return super.visitMethod(access, name, descriptor, signature, exceptions);
     }
 
     @Override
@@ -555,14 +523,14 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
         for (MethodInfo method : indexView.getClassByName(baseType.dotName()).methods()) {
             String descriptor = method.descriptor(type -> typeArguments.getOrDefault(type, OBJECT).get());
             AnnotationInstance bridge = method.annotation(DOTNAME_GENERATE_BRIDGE);
-            if (!definedMethods.containsKey(method.name() + descriptor) && bridge != null) {
+            if (!userMethods.contains(method.name() + "/" + descriptor) && bridge != null) {
                 generate(method);
                 if (needsJvmBridge(method)) {
                     String bridgeDescriptor = bridgeMethodDescriptor(method, type -> {
                         ByteCodeType mapped = typeArguments.get(type);
                         return mapped != null ? mapped.get() : null;
                     });
-                    if (!definedMethods.containsKey(method.name() + bridgeDescriptor)) {
+                    if (!userMethods.contains(method.name() + "/" + bridgeDescriptor)) {
                         generateBridge(method, bridgeDescriptor);
                     }
 

